### PR TITLE
Issue-2748 - Add ByteBuffer Hex init & write

### PR DESCRIPTION
**Pull request by ali-ahsan-ali**
Add ability for below two functions 
`ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")`
` buffer.writeHexEncodedBytes("68656c6c6f20776f726c640a")`

### Motivation:

https://github.com/apple/swift-nio/issues/2748

### CODE REVIEW

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 1]
```diff
@@ -12,8 +12,30 @@
 //
 //===----------------------------------------------------------------------===//
 
+import Foundation
```
Comment (ali-ahsan-ali) [inline Sources/NIOCore/ByteBuffer-hex.swift:15]: Is this allowed in NioCore? If not, can I use the "replaceWith" that is only MacOs 13.0 and above? Or should I implement the ability to replace white spaces of a string custom in this module?

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:15]: Ignore, found a solve.

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 2]
```diff
@@ -88,6 +88,24 @@ extension ByteBuffer {
         )
     }
 
+    // MARK: Hex encoded string APIs
+    /// Write `hex encoded string` into this `ByteBuffer`, moving the writer index forward appropriately.
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-aux.swift:92]:
```suggestion
    /// Write `string` into this `ByteBuffer` in ASCII hexadecimal, moving the writer index forward appropriately.
```

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 3]
```diff
@@ -263,3 +283,23 @@ extension ByteBuffer {
         }
     }
 }
+
+extension String {
+    /// Plain decode a string representing a hexadecimal sequence them into an UInt8 sequence
+    /// - Complexity: O(n)
+    public var hexPlainDecodedBytes: [UInt8] {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:290]: This should not be `public`.

We also don't want to round-trip it through an Array: we should be able to convert directly into `ByteBuffer`.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 4]
```diff
@@ -14,6 +14,26 @@
...
+    /// This will allocate a new `ByteBuffer` with enough space to fit the hex decoded `bytes` and potentially some extra space
+    /// using the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` we
+    ///         recommend using `channel.allocator.buffer(integer:)`. Or if you want to write multiple items into the
+    ///         buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeHexEncodedBytes` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    init(hexEncodedBytes string: String) {
+        let hexPlainDecodedBytes = string.hexPlainDecodedBytes
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:28]: Rather than use this, we can create the buffer `var` and `writeHexEncodedString` into it.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 3 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:290]: Screenshot 2024-08-13 at 11 39 23 am

Without turning it into an array we lose 2 things 

1. The ability to count the array for the exact number of bytes. The underestimated count can give 0 for sequence length that that are really 11 and it then means either I cannot allocate a buffer of the correct length ( or return early without setting any bytes). 

So if in the initliser someone tried to enter a hex string without anything in it, it would still return an empty buffer, but after some extrannuous computation and going through the setBytes slow path.

Or again, if someone tried to write empty bytes.

2. It always uses the slow method for setting bytes in the sequence rather than the contiuous if available which is done always in an array from my testing. 

Is this extra rount trip from sequence to array worth more than the cost of the above? It is becoming a difficult question for me to answer. After your comment I am leaning towards it is not worth the extra cost, seems like there is a change it takes the slow path ANYWAYS with an array. So the fast case might be faster, but the slow case slower. 

Putting my brain dump here for challenging the outcome or the thought process.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 4 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:28]: Can I challenge this? It doesnt seem like the pattern.
Screenshot 2024-08-13 at 11 58 16 am

Would this suffice or do you want me to create that var? 
Screenshot 2024-08-13 at 11 58 32 am

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:28]: Is it because I am calling that extra computer variable inside the initialiser that makes this not part of the original pattern and needings something new.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 3 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift]: Made the changes. Take a look when you find the time

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 4 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift]: Made the changes, but I think my challenge still stands.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 5]
```diff
@@ -14,6 +14,22 @@
...
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space to fit the hex decoded `bytes` and potentially some extra space
+    /// using the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` we
+    ///         recommend using `channel.allocator.buffer(integer:)`. Or if you want to write multiple items into the
+    ///         buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeHexEncodedBytes` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    init(hexEncodedBytes string: String) {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:27]: You right pointed out that we need to move this to `ByteBufferAllocator`, and then have this method call that one.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 6]
```diff
@@ -263,3 +279,24 @@ extension ByteBuffer {
...
     }
 }
+
+extension String {
+    /// Plain decode a string representing a hexadecimal sequence them into an UInt8 sequence
+    /// - Complexity: O(n)
+    @inlinable
+    var hexPlainDecodedBytes: UnfoldSequence<UInt8, String> {
+        let stringWithoutWhiteSpaces = self.filter { !$0.isWhitespace }
+        return sequence(
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:289]: I think this might be a bit too clever, and I wonder if we should try a simpler pattern. Instead of doing a `filter` followed by a `sequence`, all working on the `String`, can I propose we do a for loop? Something like this:

```swift
extension UInt8 {
    fileprivate var isASCIIWhitespace: Bool {
        switch self {
        case UInt8(ascii: "\n"), UInt8(ascii: "\t"), UInt8(ascii: "\n"), UInt8(ascii: "\r"):
            return true
        default: return false
        }
    }

    var asciiHexNibble: UInt8? {
        switch self {
        case UInt8(ascii: "0")...UInt8(ascii: "9"):
            return self - UInt8(ascii: "0")
        case UInt8(ascii: "a")...UInt8(ascii: "f"):
            return self - UInt8(ascii: "a")
        case UInt8(ascii: "A")...UInt8(ascii: "F"):
            return self - UInt8(ascii: "A")
        default:
            return nil
        }
    }
}

extension Substring.UTF8View {
    mutating func popNextHexByte() -> UInt8? {
        while let nextByte = self.first, nextByte.isASCIIWhitespace {
            self = self.dropFirst()
        }
        
        guard let firstHex = self.popFirst()?.asciiHexNibble, 
            let secondHex = self.popFirst()?.asciiHexNibble else {
            return nil
        }
        
        return (firstHex << 4) | secondHex
    }
}

extension ByteBuffer {
    mutating func writeHexEncodedBytes(_ hexString: String) -> Int {
        var slice = hexString.utf8[...]
        var written = 0
        while let nextByte = slice.popNextHexByte() {
            written += self.writeInteger(nextByte)
        }
        return written
    }
}
```

Comment (Lukasa) [reply Sources/NIOCore/ByteBuffer-hex.swift:289]: As a side note, we should probably amend the above to `throw` if the input string is invalid.

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:287]: I cannot wrap my head around the ascii nibble and popNextHexByte function. It seems to be decoding the letters wrong. I still feel like turning the String into Hex using Uint(string, radix: 16) is the simple and correct way to do things. 

If the sequence is over complicating it, I can implement it using a for loop. Something like the below. 

```swift
enum HexDecodingError: Error {
    case invalidHexLength
}

extension Substring {
    mutating func popNextHexByte() throws -> UInt8? {
        while let next = self.first, next.isWhitespace {
            self = self.dropFirst()
        }
        
        guard let firstHex = self.popFirst() else {
            return nil
        }
        
        while let next = self.first, next.isWhitespace {
            self = self.dropFirst()
        }
        
        guard let secondHex = self.popFirst() else {
            throw HexDecodingError.invalidHexLength
        }
        
        return UInt8(String(firstHex) + String(secondHex), radix: 16)
    }
}

extension ByteBuffer {
    @discardableResult
    public mutating func writeHexEncodedBytes(_ hexString: String) throws -> Int {
        var slice = hexString[...]
        var written = 0
        while let nextByte = try slice.popNextHexByte() {
            written += self.writeInteger(nextByte)
        }
        return written
    }
}
```

File: Tests/NIOCoreTests/ByteBufferTest.swift [hunk 7]
```diff
@@ -1878,6 +1889,29 @@ class ByteBufferTest: XCTestCase {
         XCTAssertEqual(expected, actual)
     }
 
+    func testWriteHexEncodedBytes() throws {
+        var buffer = ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")
```
Comment (weissi) [inline Tests/NIOCoreTests/ByteBufferTest.swift:1893]: This doesn't `throw` and is non-`Optional`, what happens if the input isn't a valid hex string? Examples

- `686`
- `686hello`
- `!.=/`
- `1 2 3 4 5 6 7 8` (this one could be argued is valid as `12 34 56 78`

@Lukasa WDYT is the right shape of the API here?

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 8]
```diff
@@ -756,6 +769,18 @@ extension ByteBufferAllocator {
         return buffer
     }
 
+    /// Create a fresh `ByteBuffer` containing the `bytes` decoded from the ASCII `hexEncodedBytes` string .
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space to fit `bytes` and potentially some extra space.
+    ///
+    /// - returns: The `ByteBuffer` containing the written bytes.
+    @inlinable
+    public func buffer(hexEncodedBytes string: String) -> ByteBuffer {
```
Comment (weissi) [inline Sources/NIOCore/ByteBuffer-aux.swift:778]: This is an uncommon API shape in NIO, this should be spelled as an `init`ialiser, no?

File: Tests/NIOCoreTests/ByteBufferTest.swift (hunk 7 above)
Comment (ali-ahsan-ali) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1893]: Once my above comment about how to implement it is answered. I wil add throws for these edge cases.

File: Sources/NIOCore/ByteBuffer-aux.swift (hunk 8 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:778]: String, substring all have the same methods in ByteBufferAllocator. I think the initialiser you are looking for is below. 
I think this is following the pattern of the repo. 

Screenshot 2024-08-19 at 1 29 41 PM

File: Tests/NIOCoreTests/ByteBufferTest.swift (hunk 7 above)
Comment (Lukasa) [reply Tests/NIOCoreTests/ByteBufferTest.swift:1893]: I suspect we should throw.

File: Sources/NIOCore/ByteBuffer-aux.swift (hunk 8 above)
Comment (Lukasa) [reply Sources/NIOCore/ByteBuffer-aux.swift:778]: Yeah, this is the standard shape for ByteBufferAllocator. While I agree with you that in hindsight this was the _wrong_ API shape, it's where we are now and the consistency does matter.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 6 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:287]:
```swift 
func testWriteHexEncodedBytes() throws {
        var buffer = ByteBuffer(hexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")
        XCTAssertEqual(try buffer.writeHexEncodedBytes("68656c6c6f20776f726c64"), 11)
        XCTAssertEqual(try buffer.writeHexEncodedBytes("     0a    "), 1)
        XCTAssertEqual(try buffer.writeHexEncodedBytes(""), 0)
        XCTAssertEqual(try buffer.writeHexEncodedBytes("      "), 0)
        XCTAssertThrowsError(try buffer.writeHexEncodedBytes("    1  "))
        print(buffer.hexDump(format: .plain))
        XCTAssertEqual(ByteBuffer(string: "hello world\nhello world\n"), buffer)
    }
```

Testing hello world, it seems to get caught up on anything that goes through the below case. There seems to just be something wrong with that calculation. I will try understand the calculation more and solve it. 

Screenshot 2024-08-18 at 12 48 00 PM

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:287]: I founda  fix. Will push something using the implementation you suggested.

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 9]
```diff
@@ -88,6 +88,24 @@ extension ByteBuffer {
...
+    /// - parameters:
+    ///     - plainHexEncodedBytes: The hex encoded string to write.
+    /// - returns: The number of bytes written.
+    @discardableResult
+    @inlinable
+    public mutating func writePlainHexEncodedBytes(_ plainHexEncodedBytes: String) throws -> Int {
+        var slice = plainHexEncodedBytes.utf8[...]
+        var written = 0
+
+        while let nextByte = try slice.popNextHexByte() {
```
Comment (ali-ahsan-ali) [inline Sources/NIOCore/ByteBuffer-aux.swift:102]: Will a throw here shut down the channel and mean the inconsistent state of the byte buffer with only half the values written is OK? 

OR Should I do an initial pass and validate BEFORE then going on and writing them to the buffer to ensure a consistent state? 

I did it initially the latter way, but poking around I assumed the channel closes and that bytebuffer is no longer used.

File: Tests/NIOCoreTests/ByteBufferTest.swift (hunk 7 above)
Comment (ali-ahsan-ali) [reply Tests/NIOCoreTests/ByteBufferTest.swift]: Can you let me know if it looks OK. I have made everything throw including the initialiser and the write method.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 10]
```diff
@@ -263,3 +277,53 @@ extension ByteBuffer {
...
 }
+
+enum HexDecodingError: Error {
+    case invalidHexLength
+}
+
+extension UInt8 {
+    fileprivate var isASCIIWhitespace: Bool {
+        switch self {
+        case UInt8(ascii: "\n"), UInt8(ascii: "\t"), UInt8(ascii: "\n"), UInt8(ascii: "\r"), UInt8(ascii: " "):
```
Comment (ali-ahsan-ali) [inline Sources/NIOCore/ByteBuffer-hex.swift:288]: static let?

File: Tests/NIOCoreTests/ByteBufferTest.swift [hunk 11]
```diff
@@ -1878,6 +1889,35 @@ class ByteBufferTest: XCTestCase {
...
+    func testWriteHexEncodedBytes() throws {
+        var buffer = try ByteBuffer(plainHexEncodedBytes: "68 65 6c 6c 6f 20 77 6f 72 6c 64 0a")
+        XCTAssertEqual(try buffer.writePlainHexEncodedBytes("68656c6c6f20776f726c64"), 11)
+        XCTAssertEqual(try buffer.writePlainHexEncodedBytes("     0a    "), 1)
+        XCTAssertEqual(try buffer.writePlainHexEncodedBytes(""), 0)
+        XCTAssertEqual(try buffer.writePlainHexEncodedBytes("      "), 0)
+        XCTAssertThrowsError(try buffer.writePlainHexEncodedBytes("    1  "))
+        XCTAssertThrowsError(try buffer.writePlainHexEncodedBytes("    1"))
+        XCTAssertThrowsError(try buffer.writePlainHexEncodedBytes("1       "))
+        // The first byte (68 = "h") is valid, the method throws and the valid byte IS written to the ByteBuffer
```
Comment (ali-ahsan-ali) [inline Tests/NIOCoreTests/ByteBufferTest.swift:1901]: This feels wrong

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 6 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift]: @Lukasa

**Comment by ali-ahsan-ali** [comment]
@weissi please take a look when you find the time

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 12]
```diff
@@ -88,6 +88,30 @@ extension ByteBuffer {
         )
     }
 
+    // MARK: Hex encoded string APIs
+    /// Write `string` into this `ByteBuffer` in ASCII hexadecimal, moving the writer index forward appropriately.
+    /// - parameters:
+    ///     - plainHexEncodedBytes: The hex encoded string to write.
+    /// - returns: The number of bytes written.
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-aux.swift:95]: Can this doc comment be updated to have an example of what this format looks like?

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 13]
```diff
@@ -88,6 +88,30 @@ extension ByteBuffer {
...
+        var bytesToWrite: [UInt8] = []
+
+        // Single pass to ensure string is a valid representation of hex bytes
+        while let nextByte = try slice.popNextHexByte() {
+            bytesToWrite.append(nextByte)
+        }
+
+        for byte in bytesToWrite {
+            written += self.writeInteger(byte)
+        }
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-aux.swift:110]: We don't need the intermediate array. Instead of using `bytesToWrite` and calling `append`, just call `writeInteger` directly.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 14]
```diff
@@ -14,6 +14,20 @@
...
+    ///
+    /// This will allocate a new `ByteBuffer` with enough space to fit the hex decoded `bytes` and potentially some extra space
+    /// using the default allocator.
+    ///
+    /// - info: If you have access to a `Channel`, `ChannelHandlerContext`, or `ByteBufferAllocator` we
+    ///         recommend using `channel.allocator.buffer(integer:)`. Or if you want to write multiple items into the
+    ///         buffer use `channel.allocator.buffer(capacity: ...)` to allocate a `ByteBuffer` of the right
+    ///         size followed by a `writeHexEncodedBytes` instead of using this method. This allows SwiftNIO to do
+    ///         accounting and optimisations of resources acquired for operations on a given `Channel` in the future.
+    init(plainHexEncodedBytes string: String) throws {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:27]: This should be `public`.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 15]
```diff
@@ -341,3 +355,54 @@ extension ByteBuffer {
         }
     }
 }
+
+enum HexDecodingError: Error {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:359]: Can we make this `public`, and make it a `struct`?

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 16]
```diff
@@ -341,3 +355,54 @@ extension ByteBuffer {
...
+    case invalidHexLength
+    case invalidCharacter
+}
+
+extension UInt8 {
+    fileprivate var isASCIIWhitespace: Bool {
+        [UInt8(ascii: "\n"), UInt8(ascii: "\t"), UInt8(ascii: "\n"), UInt8(ascii: "\r"), UInt8(ascii: " ")].contains(
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:366]: We don't need the duplicate.

```suggestion
        [UInt8(ascii: "\t"), UInt8(ascii: "\n"), UInt8(ascii: "\r"), UInt8(ascii: " ")].contains(
```

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 17]
```diff
@@ -341,3 +355,54 @@ extension ByteBuffer {
...
+            self
+        )
+    }
+
+    fileprivate var asciiHexNibble: UInt8? {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:371]: Never returns nil.

```suggestion
    fileprivate var asciiHexNibble: UInt8 {
```

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 18]
```diff
@@ -341,3 +355,54 @@ extension ByteBuffer {
...
+    mutating func popNextHexByte() throws -> UInt8? {
+        while let nextByte = self.first, nextByte.isASCIIWhitespace {
+            self = self.dropFirst()
+        }
+
+        guard let firstHex = try self.popFirst()?.asciiHexNibble else {
+            return nil  // No next byte to pop
+        }
+
+        while let nextByte = self.first, nextByte.isASCIIWhitespace {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:398]: I don't think we allow whitespace between the two hex characters.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 15 above)
Comment (weissi) [reply Sources/NIOCore/ByteBuffer-hex.swift:359]: Also needs to _either_

- be nested into `ByteBuffer`
- or have a `NIO` prefix as per the public API guarantees that we don't add top-level things that aren't `NIO`-prefixed

File: Sources/NIOCore/ByteBuffer-aux.swift (hunk 13 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:110]: XCTAssertThrowsError(try buffer.writePlainHexEncodedBytes("68 1"))

If I write it the way you have suggested, this method will throw, BUT, the 68 will be written to the byte. Would that not leave the byte in an inconsistent or unintended state potentially leading to some confusion?

Comment (Lukasa) [reply Sources/NIOCore/ByteBuffer-aux.swift:110]: We can save the original writer index and restore it when we throw the error.

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:110]: Thanks, will do

File: Sources/NIOCore/ByteBuffer-aux.swift (hunk 12 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:99]: LMK if latest push meets expectation.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 19]
```diff
@@ -340,4 +354,55 @@ extension ByteBuffer {
             }
         }
     }
+
+    public struct HexDecodingError: Error {
+        let kind: HexDecodingErrorKind
+
+        public enum HexDecodingErrorKind {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:361]: This `enum` needs to be private, or we can't safely use this type.

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 20]
```diff
@@ -340,4 +354,55 @@ extension ByteBuffer {
             }
         }
     }
+
+    public struct HexDecodingError: Error {
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:358]: Can you add a docstring here?

**Review by Lukasa — COMMENTED** [review]
Ok awesome, this is getting a _lot_ closer. Only a few little notes here.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 19 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:361]: Are you suggesting internal? Private would mean this would be impossible (to my understanding) to initialise (atleast the way I have done so).

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 21]
```diff
@@ -91,6 +91,29 @@ extension ByteBuffer {
...
+    @inlinable
+    public mutating func writePlainHexEncodedBytes(_ plainHexEncodedBytes: String) throws -> Int {
+        var slice = plainHexEncodedBytes.utf8[...]
+        let initialWriterIndex = self._writerIndex
+
+        do {
+            while let nextByte = try slice.popNextHexByte() {
+                self.writeInteger(nextByte)
+            }
+            return Int(self._writerIndex - initialWriterIndex)
```
[the remark this answers is not in the document]
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:110]: Easy.

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 22]
```diff
@@ -91,6 +91,29 @@ extension ByteBuffer {
...
+        } catch {
+            self._moveWriterIndex(to: initialWriterIndex)
```
[the remark this answers is not in the document]
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-aux.swift:112]: Easy.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 20 above)
Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift:359]: Updated. Take a look when you are free.

File: Sources/NIOCore/ByteBuffer-hex.swift (hunk 19 above)
Comment (Lukasa) [reply Sources/NIOCore/ByteBuffer-hex.swift]: Private should be fine, you'd use `public static let`s on the parent to access the fields:

```swift
public struct HexDecodingError: Error {
    private let kind: Kind

    enum Kind {
        case invalidHexLength
        case invalidCharacter
    }
    
    public static let invalidHexLength = HexDecodingError(kind: .invalidHexLength)
    public static let invalidCharacter = HexDecodingError(kind: .invalidCharacter)
}
```

Comment (ali-ahsan-ali) [reply Sources/NIOCore/ByteBuffer-hex.swift]: thanks will find time to implement.

File: Sources/NIOCore/ByteBuffer-aux.swift [hunk 23]
```diff
@@ -91,6 +91,29 @@ extension ByteBuffer {
         )
     }
 
+    // MARK: Hex encoded string APIs
+    /// Write `string` into this `ByteBuffer` in ASCII hexadecimal, moving the writer index forward appropriately. This method will throw if the string input is not of the "plain" hex encoded format.
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-aux.swift:95]:
```suggestion
    /// Write ASCII hexadecimal `string` into this `ByteBuffer` as raw bytes, decoding the hexadecimal & moving the writer index forward appropriately. This method will throw if the string input is not of the "plain" hex encoded format.
```

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 24]
```diff
@@ -14,6 +14,20 @@
 
 extension ByteBuffer {
 
+    /// Create a fresh `ByteBuffer` containing the `bytes` decoded from  the string representation of `plainHexEncodedBytes`.
```
Comment (Lukasa) [inline Sources/NIOCore/ByteBuffer-hex.swift:17]:
```suggestion
    /// Create a fresh `ByteBuffer` containing the `bytes` decoded from the string representation of `plainHexEncodedBytes`.
```

**Review by Lukasa — COMMENTED** [review]
Fantastic, this is looking really good. A pair of very minor textual tweaks and then I'm happy. Thanks for sticking with this!

File: Sources/NIOCore/ByteBuffer-hex.swift [hunk 25]
```diff
@@ -340,4 +354,58 @@ extension ByteBuffer {
             }
         }
     }
+
+    /// An error that is thrown when an invalid hex encoded string was attempted to be written to a ByteBuffer.
+    public struct HexDecodingError: Error {
+        let kind: HexDecodingErrorKind
+
+        enum HexDecodingErrorKind {
+            /// The hex encoded string was not of the expected even length.
```
Comment (ali-ahsan-ali) [inline Sources/NIOCore/ByteBuffer-hex.swift:363]: Make that change here that I never made

**Comment by ali-ahsan-ali** [comment]
> Fantastic, this is looking really good. A pair of very minor textual tweaks and then I'm happy. Thanks for sticking with this!

Doing a bit of travelling. Actually flying to London for the swift conference. Will find time to push the changes.

**Review by Lukasa — APPROVED** [review]
Awesome, I'm happy with this! Well done getting through the gauntlet again @ali-ahsan-ali, it's really good that you were able to stick with it. Much appreciated ✨ .
